packages/compass-user-data/src/user-data.ts

```diff
@@ -0,0 +1,185 @@
+    subdir,
+    basePath,
+    serialize = (content: T) => JSON.stringify(content, null, 2),
+    deserialize = JSON.parse,
+  }: UserDataOptions<T>) {
+    this.subdir = subdir;
+    this.basePath = basePath;
+    this.deserialize = deserialize;
+    this.serialize = serialize;
+  }
+
+  private async getEnsuredBasePath(): Promise<string> {
+    const basepath = this.basePath ? this.basePath : getStoragePath();
+
+    const root = path.join(basepath, this.subdir);
+
+    await fs.mkdir(root, { recursive: true });
+
+    return root;
+  }
+
+  private async getFileAbsolutePath(filepath?: string): Promise<string> {
+    const root = await this.getEnsuredBasePath();
+    const pathRelativeToRoot = path.relative(
+      root,
+      path.join(root, filepath ?? '')
+    );
+
+    if (
+      pathRelativeToRoot.startsWith('..') ||
+      path.isAbsolute(pathRelativeToRoot)
+    ) {
+      throw new Error(
+        `Invalid file path: '${filepath}' is not a subpath of ${root}.`
+      );
+    }
+
+    return path.resolve(root, pathRelativeToRoot);
+  }
+
+  private async readAndParseFile(absolutePath: string, options: ReadOptions) {
+    let data: string;
+    try {
+      data = await fs.readFile(absolutePath, 'utf-8');
+    } catch (error) {
+      log.error(mongoLogId(1_001_000_234), 'Filesystem', 'Error reading file', {
+        path: absolutePath,
+        error: (error as Error).message,
+      });
+      if (options.ignoreErrors) {
+        return undefined;
+      }
+      throw error;
+    }
+
+    try {
+      return this.deserialize(data);
+    } catch (error) {
+      log.error(mongoLogId(1_001_000_235), 'Filesystem', 'Error parsing data', {
+        path: absolutePath,
+        error: (error as Error).message,
+      });
+      if (options.ignoreErrors) {
+        return undefined;
+      }
+      throw error;
+    }
+  }
+
+  async readAll(
+    options: ReadOptions = {
+      ignoreErrors: true,
+    }
+  ) {
+    const absolutePath = await this.getFileAbsolutePath();
+    const filePathList = await fs.readdir(absolutePath);
+    const data = await Promise.allSettled(
+      filePathList.map((x) => this.readOne(x, options))
+    );
+
+    const result = {
+      data: [],
+      errors: [],
+    } as {
+      data: T[];
+      errors: Error[];
+    };
+
+    for (const item of data) {
+      if (item.status === 'fulfilled' && item.value) {
+        result.data.push(item.value);
+      }
+      if (item.status === 'rejected') {
+        result.errors.push(item.reason);
+      }
+    }
+
+    return result;
+  }
+
+  async readOne(
+    filepath: string,
+    options?: { ignoreErrors: false }
+  ): Promise<T>;
+  async readOne(
+    filepath: string,
+    options?: { ignoreErrors: true }
+  ): Promise<T | undefined>;
+  async readOne(
+    filepath: string,
+    options?: ReadOptions
+  ): Promise<T | undefined>;
+  async readOne(
+    filepath: string,
+    options: ReadOptions = {
+      ignoreErrors: true,
+    }
+  ) {
```

Comment:
So while the types here are written in a way that will suggest that returned value is matching provided generic, nothing is actually guaranteeing this, which was a thing with ampersand models using storage mixin, ampersand will remove properties that are not defined on the model and validate types on properties that are defined as props. I see two ways of addressing this:

- `UserData` class handles this somehow. For example, generic type is not just provided by the user with no guarantees, but derived from some required schema that can be used for actual runtime validation of what is being loaded from disk
- `UserData` class is not responsible for this, but then the types should not assert what they can't guarantee and readOne / readAll should return `unknown` type instead